django/core/handlers/wsgi.py

```diff
@@ -12,56 +12,46 @@
 _slashes_re = _lazy_re_compile(rb"/+")
 
 
-class LimitedStream:
-    """Wrap another stream to disallow reading it past a number of bytes."""
+class LimitedStream(IOBase):
+    """
+    Wrap another stream to disallow reading it past a number of bytes.
+
+    Based on the implementation from werkzeug.wsgi.LimitedStream
+    See https://github.com/pallets/werkzeug/blob/dbf78f67/src/werkzeug/wsgi.py#L828
+    """
 
     def __init__(self, stream, limit):
-        self.stream = stream
-        self.remaining = limit
-        self.buffer = b""
-
-    def _read_limited(self, size=None):
-        if size is None or size > self.remaining:
-            size = self.remaining
-        if size == 0:
+        self._read = stream.read
+        self._readline = stream.readline
+        self._pos = 0
+        self.limit = limit
+
+    def read(self, size=-1, /):
+        _pos = self._pos
+        limit = self.limit
+        if _pos >= limit:
             return b""
-        result = self.stream.read(size)
-        self.remaining -= len(result)
-        return result
-
-    def read(self, size=None):
-        if size is None:
-            result = self.buffer + self._read_limited()
-            self.buffer = b""
-        elif size < len(self.buffer):
-            result = self.buffer[:size]
-            self.buffer = self.buffer[size:]
-        else:  # size >= len(self.buffer)
-            result = self.buffer + self._read_limited(size - len(self.buffer))
-            self.buffer = b""
-        return result
-
-    def readline(self, size=None):
-        while b"\n" not in self.buffer and (size is None or len(self.buffer) < size):
-            if size:
-                # since size is not None here, len(self.buffer) < size
-                chunk = self._read_limited(size - len(self.buffer))
-            else:
-                chunk = self._read_limited()
-            if not chunk:
-                break
-            self.buffer += chunk
-        sio = BytesIO(self.buffer)
-        if size:
-            line = sio.readline(size)
+        if size == -1 or size is None:
+            size = limit - _pos
         else:
-            line = sio.readline()
-        self.buffer = sio.read()
+            size = min(size, limit - _pos)
+        data = self._read(size)
+        self._pos += len(data)
+        return data
+
+    def readline(self, size=-1, /):
+        _pos = self._pos
+        limit = self.limit
+        if _pos >= limit:
+            return b""
+        if size == -1 or size is None:
+            size = limit - _pos
+        else:
+            size = min(size, limit - _pos)
+        line = self._readline(size)
```

Comment:
This is not valid according to the spec:
> Servers should support the optional “size” argument to readline(), but as in WSGI 1.0, they are allowed to omit support for it.

It apparently hasn't been an issue for us in the past so I guess we can keep it like that?

Reply:
I think for this we should assume that the `size` argument is supported as, although the spec allows it to be optional, it seems to always be available from my experience. In practice most implementations are likely using interfaces from `io` which provides support.

We could catch `TypeError` and retry the call without the `size` argument for rare cases where it might not be supported? Let me know if you thing it's worth adding this...

I really dislike these sorts of things in specifications - it just makes things less robust 😞

Reply:
I am tempted to fix this once it becomes a problem and leave it like this for now.

Reply:
See also this note https://github.com/pallets/werkzeug/blob/dbf78f674dda529e0ee5cb0fad7892a04e61bec1/src/werkzeug/wsgi.py#L842-L847